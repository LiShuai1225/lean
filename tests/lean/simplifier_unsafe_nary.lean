open tactic

universe l
constants (A : Type.{l}) (op : A → A → A → A)
          (op_assoc : Π a, is_associative (op a))
          (a b x y z : A)

attribute op_assoc [instance]

print "Safe version, should not simplify"
set_option simplify.unsafe_nary false
example : (op a) ((op b) x y) z = (op a) x ((op b) y z) := by simp

print "Unsafe version, should simplify and the proof macro should be incorrect"

set_option simplify.unsafe_nary true
example : (op a) ((op b) x y) z = (op a) x ((op b) y z) := by simp
