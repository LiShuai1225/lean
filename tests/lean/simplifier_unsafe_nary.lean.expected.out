Safe version, should not simplify
simplifier_unsafe_nary.lean:12:59: error: simp tactic failed to simplify
state:
⊢ op a (op b x y) z = op a x (op b y z)
simplifier_unsafe_nary.lean:12:59: error: failed to add declaration 'example' to environment, value has metavariables
remark: set 'formatter.hide_full_terms' to false to see the complete term
  ?M_1
Unsafe version, should simplify and the proof should be incorrect
simplifier_unsafe_nary.lean:17:0: error: type mismatch at application
  (λ a a_1 e_1 a_2 a_3, congr (congr_arg eq e_1)) (op a (op b x y) z) (op a x (op a y z))
    [flat (is_associative.op_assoc (op a)) (op a (op a x y) z = op a x (op a y z))]
term
  [flat (is_associative.op_assoc (op a)) (op a (op a x y) z = op a x (op a y z))]
has type
  op a (op a x y) z = op a x (op a y z)
but is expected to have type
  op a (op b x y) z = op a x (op a y z)
